apps/google-analytics-4/frontend/src/apis/fetchApi.ts

```diff
@@ -0,0 +1,99 @@
+import { z } from 'zod';
+import fetchWithSignedRequest from '../helpers/signed-requests';
+import { PlainClientAPI } from 'contentful-management';
+import { ApiError } from 'apis/api';
+import { Headers, ApiErrorType, ERROR_TYPE_MAP } from 'apis/apiTypes';
+
+export async function fetchFromApi<T>(
+  apiUrl: URL,
+  schema: z.ZodTypeAny,
+  appDefinitionId: string,
+  cma: PlainClientAPI,
+  headers: Headers = {}
+): Promise<T> {
+  const response = await fetchResponse(apiUrl, appDefinitionId, cma, headers);
+  const responseJson = await jsonFromResponse(response);
+  validateResponseStatus(response, responseJson);
+  parseResponseJson(responseJson, schema);
+  return responseJson;
+}
+
+async function fetchResponse(
+  url: URL,
+  appDefinitionId: string,
+  cma: PlainClientAPI,
+  headers: Headers
+): Promise<Response> {
+  try {
+    const signedResponse = await fetchWithSignedRequest(url, appDefinitionId, cma, 'GET', headers);
+    return signedResponse;
+  } catch (e) {
+    if (e instanceof TypeError) {
+      const responseError: ApiErrorType = {
+        message: e.message,
+        errorType: ERROR_TYPE_MAP.failedFetch,
+        details: null,
+        status: 500,
+      };
+      throw new ApiError(responseError);
+    }
+    throw e;
+  }
+}
+
+async function jsonFromResponse(response: Response): Promise<any> {
+  try {
+    const json = await response.json();
+    return json;
+  } catch (e) {
+    if (e instanceof SyntaxError) {
+      const responseError: ApiErrorType = {
+        message: e.message,
+        errorType: ERROR_TYPE_MAP.invalidJson,
+        details: null,
+        status: 500,
+      };
+      throw new ApiError(responseError);
+    }
+    throw e;
+  }
+}
+
+function validateResponseStatus(response: Response, responseJson: any): void {
+  if (response.status >= 400) {
+    if (isApiErrorType(responseJson.errors)) {
```

Comment:
Since we have a zod schema for this, I suggest we use that instead of a custom type guard function. Cheaper and easier. I can help show how to refactor.